pkg/ring/kv/memberlist/memberlist_client.go

```diff
@@ -0,0 +1,853 @@
+// some notifications may be lost.
+//
+// Watching ends when 'f' returns false, context is done, or this client is shut down.
+//
+// Part of kv.Client interface.
+func (m *Client) WatchPrefix(ctx context.Context, prefix string, f func(string, interface{}) bool) {
+	// we use bigger buffer here, since keys are interesting and we don't want to lose them.
+	w := make(chan string, 16)
+
+	// register watcher
+	m.watchersMu.Lock()
+	m.prefixWatchers[prefix] = append(m.prefixWatchers[prefix], w)
+	m.watchersMu.Unlock()
+
+	defer func() {
+		// unregister watcher on exit
+		m.watchersMu.Lock()
+		defer m.watchersMu.Unlock()
+
+		removeWatcherChannel(prefix, w, m.prefixWatchers)
+	}()
+
+	for {
+		select {
+		case key := <-w:
+			val, err := m.get(key)
+			if err != nil {
+				level.Warn(m.logger).Log("msg", "failed to decode value while watching for changes", "key", key, "err", err)
+				continue
+			}
+
+			if !f(key, val) {
+				return
+			}
+
+		case <-m.shutdown:
+			// stop watching on shutdown
+			return
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func removeWatcherChannel(k string, w chan string, watchers map[string][]chan string) {
+	ws := watchers[k]
+	for ix, kw := range ws {
+		if kw == w {
+			ws = append(ws[:ix], ws[ix+1:]...)
+			break
+		}
+	}
+
+	if len(ws) > 0 {
+		watchers[k] = ws
+	} else {
+		delete(watchers, k)
+	}
+}
+
+func (m *Client) notifyWatchers(key string) {
+	m.watchersMu.Lock()
+	defer m.watchersMu.Unlock()
+
+	for _, kw := range m.watchers[key] {
+		select {
+		case kw <- key:
+			// notification sent.
+		default:
+			// cannot send notification to this watcher at the moment
+			// but since this is a buffered channel, it means that
+			// there is already a pending notification anyway
+		}
+	}
+
+	for p, ws := range m.prefixWatchers {
+		if strings.HasPrefix(key, p) {
+			for _, pw := range ws {
+				select {
+				case pw <- key:
+					// notification sent.
+				default:
+					c, _ := m.watchPrefixDroppedNotifications.GetMetricWithLabelValues(p)
+					if c != nil {
+						c.Inc()
+					}
+
+					level.Warn(m.logger).Log("msg", "failed to send notification to prefix watcher", "prefix", p)
+				}
+			}
+		}
+	}
+}
+
+// CAS is part of kv.Client implementation.
+//
+// Unlike other implementation, it doesn't really do Compare-And-Swap though. New values returned from function 'f'
+// are merged into the current state. If this merge doesn't result in any change (Mergeable.Merge returns nil), then
+// operation fails and is retried again. After too many failed retries, this method returns error.
+func (m *Client) CAS(ctx context.Context, key string, f func(in interface{}) (out interface{}, retry bool, err error)) error {
+	if len(key) > maxKeyLength {
+		return fmt.Errorf("key too long: %d", len(key))
+	}
+
+	sleep := false
+	for retries := 10; retries > 0; retries-- {
+		m.casAttempts.Inc()
+
+		if sleep {
+			// wait for a while before retrying... typically merge fails because timestamp hasn't
```

Comment:
This is a workaround for a problem in Ring merge function. Merging is based on timestamps with 1 second resolution. If we call 'f' function several times in a row with no sleep, it will simply report the same timestamp, and then merge will not do any change. Solution is either to make merge more complex (which I wanted to avoid), or simply sleep for a short time in a hope that next call of 'f' will put new timestamp. (Another solution would be to introduce new Timestamp to the ring, eg. with nanosecond resolution)

So basically this is a hack to help Ring merging.

Reply:
Why are we failing if there is no merge? I would assume that if there is no change, then we will be succeeding. The new value we want to set is already set....

Reply:
There is a mismatch between two functions, 'f' and Merge.

It is up to function 'f' to detect no change, and return nil in such case. If 'f' returns new value, it is trying to write it to the store.

If Merge returns no change, it only means that Merge didn't *detect* the change, but since 'f' has reported new value, we can expect that there is *some* change. By retrying several times, the hope is that Merge will be able to see it.

(Sleep part is there because we know that Ring's Merge depends on timestamps with second-resolution.)

You may also ask, why do we use Merge in CAS in the first place? There are two reasons for that: 1) it allows us to find out what the change was and broadcast only this change, 2) merging is what other gossiping members do when they receive our state. If we don't detect any change, it's most likely that they won't either, and will ignore such message.

Reply:
Does this make sense now? Can I resolve it?

Reply:
Yes please. Just mentioned that changes in merge depend on the timestamp and not the tokens?